### PR TITLE
[Demo] Query/SG: chopped queries

### PR DESCRIPTION
**Pull request by ppanyukov**
# DEMO. NOT FOR MERGE

**TL;DR:** *NOT WORTH IT* (unless I did something dumb, in which case can revisit)

---

This is a small demo of the chunked queries:

- Part of the `[8.]` item in https://github.com/thanos-io/thanos/issues/1649
- PR is for historical purposes to record the results.

**Change:**

- The change is in querier.
- Given time interval `[mint, maxt]`, chop it into smaller `1h` intervals `[mint, mint+1h], [mint+1h, mint+2h] ... [maxt-1h, maxt]`.
- Ask Store Gateway (SG) to give results for each smaller chunk.
- Assemble results in the querier.

**Purpose:**

- See if this has positive impact on memory utilisation in both Querier and SG components;
- See if query execution time is affected in any negative way.

**Results:**

- SG memory utilisation is really much better (flat line).
- Querier is way way higher though.
- Also it's quite a bit slower.
- Query used: `count({__name__=~".+"}) by (__name__)`
- See stats and charts.

```
ORIGINAL IMPLEMENTATION:

    real    1m30.330s
    user    0m0.014s
    sys     0m0.021s

    Peak MEM:
        SG  22GB
        Q   16GB


CHUNKED:

    real    2m13.879s
    user    0m0.014s
    sys     0m0.022s

    Peak MEM:
        SG  6GB
        Q   25GB

```

**Conclusion:**

*NOT WORTH IT* 

- Looks to shift the problem from SG to Querier at the cost of latency.
- Questionable benefit.
- Unless I did something dumb, in which case can revisit later.

** Chart **

Screenshot 2019-10-18 at 14 27 13

### CODE REVIEW

**Comment by ppanyukov** [comment]
Closing as it's for historical purposes only.

**Comment by bwplotka** [comment]
Nice! Good work, fun fact is that overall it is less memory (:
